sb3_contrib/crossq/crossq.py

```diff
@@ -0,0 +1,330 @@
+            # one forward pass with shape (n_critics, 2 * batch_size, 1).
+            #
+            # This has two reasons:
+            # 1. According to the paper obs/actions and next_obs/next_state_actions are differently
+            #    distributed which is the reason why "naively" applying Batch Normalization in SAC fails.
+            #    The batch statistics have to instead be calculated for the mixture distribution of obs/next_obs
+            #    and actions/next_state_actions. Otherwise, next_obs/next_state_actions are perceived as
+            #    out-of-distribution to the Batch Normalization layer, since running statistics are only polyak averaged
+            #    over from the live network and have never seen the next batch which is known to be unstable.
+            #    Without target networks, the joint forward pass is a simple solution to calculate
+            #    the joint batch statistics directly with a single forward pass.
+            #
+            # 2. From a computational perspective a single forward pass is simply more efficient than
+            #    two sequential forward passes.
+            all_obs = th.cat([replay_data.observations, replay_data.next_observations], dim=0)
+            all_acts = th.cat([replay_data.actions, next_actions], dim=0)
+
+            self.critic.set_training_mode(True)
+            all_q_values = th.cat(self.critic(all_obs, all_acts), dim=1)
+            self.critic.set_training_mode(False)
+            current_q_values, next_q_values = th.split(all_q_values, batch_size, dim=0)
+            current_q_values = th.split(current_q_values, [1, 1], dim=1)  # type: ignore
+
+            with th.no_grad():
+                # Compute the target Q value
+                next_q_values = next_q_values.detach()
+                next_q_values, _ = th.min(next_q_values, dim=1, keepdim=True)
+
+                # add entropy term
+                next_q_values = next_q_values - ent_coef * next_log_prob.reshape(-1, 1)
+                # td error + entropy term
+                target_q_values = replay_data.rewards + (1 - replay_data.dones) * self.gamma * next_q_values
+
+            # Compute critic loss
+            critic_loss = 0.5 * sum(F.mse_loss(current_q, target_q_values.detach()) for current_q in current_q_values)
+            assert isinstance(critic_loss, th.Tensor)  # for type checker
+            critic_losses.append(critic_loss.item())  # type: ignore[union-attr]
+
+            # Optimize the critic
+            self.critic.optimizer.zero_grad()
+            critic_loss.backward()
+            self.critic.optimizer.step()
+
+            # Compute actor loss
+            self.critic.set_training_mode(False)
+            q_values_pi = th.cat(self.critic(replay_data.observations, actions_pi), dim=1)
+            self.critic.set_training_mode(False)
+
+            min_qf_pi, _ = th.min(q_values_pi, dim=1, keepdim=True)
```

Comment:
no policy delay for now?
We should probably mention it in the docstring of the class + in the doc

Reply:
Good catch. I forgot adding it, since in our experiments it's not needed for sample efficiency. I added it now, since the policy delay increases the compute efficiency.